### ISSUE TITLE
Display a violation at 1:1 for each file instead of throwing an error on unrecognised rules

### ISSUE THREAD

**Issue by jamiedust** [stylelint/stylelint#3147]
Firstly I would like to say thanks for a great project, I have been introducing my team to stylelint on a couple of projects.

> Describe the issue. Is it a bug or a feature request (new rule, new option, etc.)?

Feature request - don't throw an error on unrecognised rules.

> Which rule, if any, is this issue related to?

It could be any, but as an example: `font-family-no-missing-generic-family-keyword`

> What CSS is needed to reproduce this issue?

Not applicable.

> What stylelint configuration is needed to reproduce this issue?

This should affect any configuration where a user is using an undefined rule.

> Which version of stylelint are you using?

Not really applicable, but in this case 7.x.x

> How are you running stylelint: CLI, PostCSS plugin, Node API?

CLI with `stylelint 'src/**/*.js` - I am using styled components, but this issue affects others.

> Does your issue relate to non-standard syntax (e.g. SCSS, nesting, etc.)?

No

> What did you expect to happen?

Stylelint to run correctly and lint my files.

> What actually happened (e.g. what warnings or errors you are getting)?

Stylelint fails to run with:

```shell
Error: Undefined rule font-family-no-missing-generic-family-keyword
```

The issue itself has been mentioned before, for example here #3067 - the answer is to use the correct version of stylelint, however there are times when that is not always easy, for example when using other tools which wrap stylelint.

In my case I am using Neutrino with their stylelint preset, which uses an older version of stylelint-webpack-plugin, which uses stylelint v7. I have asked the maintainers of neutrino to update their project, as I don't want to have to start locking down my version of stylelint-config-standard.

This could be avoided if instead of throwing an error and failing to run, stylelint ignored the rule, perhaps displaying a warning in the command line output?

Again, thanks for the project.

**Comment by evilebottnawi** [comment]
@jamiedust hm, looks reasonable, we can `"my-strange-rule": null` (if `null` not throw error). And you can disable some rules for each project.
@stylelint/core what do you think?

**Comment by jamiedust** [comment]
Yes I would be happy to provide a null rule in order to prevent the error.

**Comment by jeddy3** [comment]
@jamiedust Thanks for the clear write-up.

@stylelint/core Any objections to changing the behaviour to ignoring unknown rules and displaying a warning instead of throwing an error?

**Comment by CAYdenberg** [comment]
Just trying to understand, what would (for e.g.) the atom linter do with such a warning? The only downside I can see is that if for instance you misspell a rule you need the information necessary to fix it, whether using the CLI or an editor plugin etc.. 

I believe that ESLint errors on an unrecognizable configuration - maybe for this reason? Or maybe in later versions actually reports at line 1 col 1 of every file linted?

**Comment by Arcanemagus** [comment]
Since `stylelint` is often embedded into other things, I'd say it's probably a good idea to change to this, but there are a few cases to conder:

* What happens when there are _multiple_ unknown rules?
  * Probably just print them all, could get spammy though.
* What if _all_ the rules are invalid?

----

> Just trying to understand, what would (for e.g.) the atom linter do with such a warning?

Currently the `throw`n error is caught and displayed to the user using Atom's standard error notification interface.

If this was changed to a warning it would instead be displayed as a regular message to the user, encompassing the first line of the file that was open when the `lint` was requested. This style of interaction has generally seemed to work better for users, in other providers where errors are _far_ more common (I'm looking at you `jshint`...) _all_ errors are trapped this way.

tl;dr It doesn't actually matter for `linter-stylelint`, it handles either case already 😉.

**Comment by jeddy3** [comment]
> tl;dr It doesn't actually matter for linter-stylelint, it handles either case already 😉.

Good to know! :)

> I believe that ESLint errors on an unrecognizable configuration - maybe for this reason? Or maybe in later versions actually reports at line 1 col 1 of every file linted?

I can confirm this is the current behaviour of ESLint. It also prints each unknown rule.

It sounds like we're OK with the behaviour being changed. I'll label up accordingly.
